prog/types.go

```diff
@@ -718,24 +722,24 @@ func foreachTypeImpl(meta *Syscall, preorder bool, f func(t Type, ctx *TypeCtx))
 		}
 		switch a := (*ptr).(type) {
 		case *PtrType:
-			rec(&a.Elem, a.ElemDir)
+			rec(&a.Elem, a.ElemDir, optional)
 		case *ArrayType:
-			rec(&a.Elem, dir)
+			rec(&a.Elem, dir, optional)
 		case *StructType:
 			if seen[a] {
 				break // prune recursion via pointers to structs/unions
```

Comment:
I wonder whether `seen` should actually be `Type` x `Dir` x `Optional`.

If the same type is included several times and it just so happened that the first time we hit it it was `DirOut` or `Optional=true`, we just won't go in there again, while on other occasions it may actually be mandatory and `DirIn`.

Reply:
Yep, that's correct. I added a commit to test this (both the `Dir` and the `Optional` cases) with `sys/test` and verified it failed without a fix. The fix you suggested is in commit 3, the tests in commit 4. Commit 2 tests the original patch I submitted.